File: lib/generate_diagram.rb

require 'time'
require 'rdf'
require 'set'

include RDF

# Generate some visualizations

class GenerateDiagram
  # Convert an iri to a text label
  def self.uml_name(iri)
    case iri
    when Node
      iri.to_s

    else
      qn = iri.qname(prefixes: Prefixes)
      v = if qn[0] == :obo
            "\"bfo:#{iri.attributes[:label][:en]}\""
          else
            qn.join(':')
          end
      if qn.first == :core
        "core:[[#{iri} #{qn.last}]]"
      else
        v
      end
    end
  rescue
    puts "Error: #{$!}\n#{iri.inspect}"
    raise
  end

  def color(iri)
    prefix = iri.qname(prefixes: Prefixes).first
    if c = Colors[prefix]
      "##{c}"
    else
      '#DarkGrey'
    end
  end

  # Add a type associated with an iri.
  def self.add_type(iri, type)
    @@types[iri] = type
  end

  # Get the types associated with each individual
  def self.get_types(graph)
    @@types = Hash.new
    stmts = RDF::Query.execute(graph) do
      pattern [:o, RDF::RDFV.type, :type]  
    end
    stmts.each do |s|
      add_type(s.o, s.type)
    end
  end

  # Add a list of terms that will have html links
  def self.add_linked_terms(terms)
    @@linked_terms ||= Set.new
    Array(terms).each { |term| @@linked_terms << term }
  end

  def self.linked_terms
    @@linked_terms
  end

  # Initialize the object with the output filename and an array of statements
  def initialize(filename, statements)
    @filename = filename
    @statements = statements
    @objects = Hash.new

    @statements.each do |stmt|
      unless stmt.predicate == RDF::RDFV.type
        add_object(stmt.subject)
        unless RDF::Literal === stmt.object
          add_object(stmt.object)
        end
      end
    end
  end

  # Add an object associated of a short 'o<n>' where n is a monotonically increasing number
  def add_object(iri)
    unless @objects[iri]
      @objects[iri] = "o#{@objects.size + 1}"
    end
  end

  # Print an object with its type
  def print_object(iri)
    return if iri == RDF::RDFV.nil
    
    if @@linked_terms.include?(iri)
      title = "#{iri.qname.first}:[[./#{iri.qname.last}.html #{iri.qname.last}]]"
    else
      title = self.class.uml_name(iri)
    end
    if t = @@types[iri]
      @f.puts "individual(#{obj(iri)}, #{title}, #{self.class.uml_name(t)})"
    else
      @f.puts "individual(#{obj(iri)}, #{title})"
    end
  end

  # Get an object for an iri
  def obj(iri)
    @objects[iri]
  end

  # Print a statment with a subject, predicate, and object. If this is a literal, print it as part of the individual.
  def print_stmt(stmt)
    unless stmt.predicate == RDF::RDFV.type or stmt.object == RDF::RDFV.nil
      if RDF::Literal === stmt.object
        @f.puts "data(#{obj(stmt.subject)}, #{self.class.uml_name(stmt.predicate)}, #{stmt.object.value})"
      else
        @f.puts "property(#{obj(stmt.subject)}, #{self.class.uml_name(stmt.predicate)}, #{obj(stmt.object)})"
      end
    end  
  end

  # Generate the html and puml files. Then run plantuml app on the file.
  def generate
    File.open("#{@filename}.html", 'w') do |f|
      f.puts <<EOT
<!doctype html>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>#{File.basename @filename}</title>
  </head>
  <body>
    <object data="./#{File.basename @filename}.svg"/>
  </body>
</html>
EOT
    end
    
    File.open("#{@filename}.puml", 'w') do |f|
      @f = f
      f.puts <<EOT
@startuml
' !include https://raw.githubusercontent.com/iofoundry/ontopuml/refs/heads/Development/iof.iuml
!include ../iof.iuml

<style>
object {
  HyperlinkColor #C4E1E6
}
</style>

left to right direction
skinparam linetype polyline
title #{File.basename @filename}

!$default_direction = down

!$namespace_colors = { "bfo":"DFA702", 
                      "iof":"1E90FF", 
                      "ns":"Green", 
                      "ns1":"76608A",
EOT
      color = {
        core: "000000",
        ex: "Green",
        data: "76608A",
        units: "8A2D3B",
        "cmns-dsg": "3A0519",
        qp: "641B2E"
      }

      f.puts Prefixes.map { |k, v|
        "\"#{k}\":\"#{color[k]}\"" if color[k]
      }.compact.join(',')
      f.puts "}"
      
      @objects.each do |k, v|
        print_object(k)
      end
      
      @statements.each do |stmt|
        print_stmt(stmt)
      end
      
      f.puts("@enduml")
    end
    @f = nil

    puts "Exec: plantuml -tsvg #{@filename}"
    `plantuml -tsvg #{@filename}.puml`
    puts "Done"

  rescue
    puts "Error occurred during wriing of file: #{$!}"
    puts $!.backtrace.join("\n")
  end
end

File: generate_diagram.py
from ontologies import *
import os
import owlready2 as owl
import owlready2.base as ob
import subprocess

class GenerateDiagram:
  linked_terms = None
  
  def _add_object(self, obj):
    if not obj in self.objects:
      self.objects[obj] = f"o{len(self.objects)}"
  
  def _obj(self, obj):
    return self.objects.get(obj, None)
  
  def __init__(self, filename, statements, vendor):
    self.filename = filename
    self.statements = statements
    self.objects = dict()
    self.vendor = vendor
    
    self.namespaces = {
      'https://spec.industrialontologies.org/ontology/core/Core/': 'core',
      'http://purl.obolibrary.org/obo/': 'bfo',
      'http://qudt.org/vocab/unit/': 'units',
      'https://www.omg.org/spec/Commons/Designators/': 'des',
      'https://spec.industrialontologies.org/ontology/qualities/Qualities/': 'qual',
      'https://spec.industrialontologies.org/ontology/qualities/Qualities-Physical/': 'qualphys',
      'http://example.org/data/': 'data',
      f"http://example.org/{vendor.name}/": 'vendor',
      "http://example.org/ontology/": 'ex',
      "http://www.w3.org/2002/07/owl#": 'owl',
      "http://www.w3.org/1999/02/22-rdf-syntax-ns#": 'rdf',
      "http://www.w3.org/2001/XMLSchema#": 'xsd',
      "http://www.w3.org/2000/01/rdf-schema#": 'rdfs'
    }
    
    for s, p, o in statements:
      if p != ob.rdf_type:
        self._add_object(s)
        if not isinstance(o, (str, int, float, owl.locstr)): self._add_object(o)
        
  def _name(self, o):
    if isinstance(o, (float, str, owl.locstr)):
      return o
    
    name = None
    ns = None
    if isinstance(o, (int)):
      iri = owl.default_world._unabbreviate(o)
      if not iri:
        return o
      
      if '#' in iri:
        ns, name = iri.rsplit('#', 1)
        ns += '#'
      else:
        ns, name = iri.rsplit('/', 1)
        ns += '/'
    else:      
      ns = o.namespace.base_iri
      if o.namespace.base_iri == BFO.base_iri:
        name = o._python_name
      else:
        name = o.name

    pre = self.namespaces.get(ns, "_")        
    return f"{pre}:{name}"
  
  def _print_object(self, f, o):
    if o == ob.rdf_nil: return
    
    if o in self.linked_terms:
      ns, name = self._name(o).split(':')
      title = f"{ns}:[[./{name}.html {name}]]"
    else:
      title = self._name(o)

    if o.is_a[0].namespace == self.vendor:
      ns, name = self._name(o.is_a[0]).split(':')
      cls = f"{ns}:[[../{self.vendor.name}.html#{name} {name}]]"
      f.write(f"individual({self._obj(o)}, {title}, {cls})\n")      
    elif o.is_a[0].namespace == Example:
      ns, name = self._name(o.is_a[0]).split(':')
      cls = f"{ns}:[[../Example.html#{name} {name}]]"
      f.write(f"individual({self._obj(o)}, {title}, {cls})\n")      
    elif o.is_a[0].namespace == Core:
      ns, name = self._name(o.is_a[0]).split(':')
      cls = f"{ns}:[[{o.is_a[0].iri} {name}]]"
      f.write(f"individual({self._obj(o)}, {title}, {cls})\n")      
    else:
      f.write(f"individual({self._obj(o)}, {title}, {self._name(o.is_a[0])})\n")    

  def _print_statement(self, f, s, p, o):
    if p != ob.rdf_type and o != ob.rdf_nil:
      if isinstance(o, (int, str, float)):
        f.write(f"data({self._obj(s)}, \"{self._name(p)}\", \"{self._name(o)}\")\n")
      else:
        f.write(f"property({self._obj(s)}, \"{self._name(p)}\", {self._obj(o)})\n")
  
  def _generate_html(self):
    with open(f"diagrams/{self.filename}.html", 'w') as f:
      f.write(f"""
<!doctype html>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>{os.path.basename(self.filename)}</title>
  </head>
  <body>
    <object data="./{os.path.basename(self.filename)}.svg"/>
  </body>
</html>
""")
  
  def _generate_diagram(self):
    with open(f"diagrams/{self.filename}.puml", 'w') as f:
      f.write(f"""
@startuml
' !include https://raw.githubusercontent.com/iofoundry/ontopuml/refs/heads/Development/iof.iuml
!include ../iof.iuml

<style>
object {{
  HyperlinkColor #C4E1E6
}}
</style>

left to right direction
skinparam linetype polyline
title {os.path.basename(self.filename)}

!$default_direction = down

!$namespace_colors = {{ "bfo":"DFA702", 
                      "iof":"1E90FF", 
                      "ns":"Green", 
                      "ns1":"76608A",
                      "core":"000000",
                      "ex":"Green",
                      "data":"76608A",
                      "vendor":"76608A",
                      "units":"8A2D3B",
                      "cmns-dsg":"3A0519",
                      "qp":"641B2E"
      }}
""")
      
      for o in self.objects:
        self._print_object(f, o)
        
      for s, p, o in self.statements:
        self._print_statement(f, s, p, o)
        
      f.write("@enduml\n")
  
  def generate(self):
    self._generate_html()
    self._generate_diagram()
    
    subprocess.run(['plantuml', '-tsvg', f"diagrams/{self.filename}.puml"])
        
      
  
      
